### ISSUE TITLE
git lfs cannot be installed when using captainhook

### ISSUE THREAD

**Issue by amenk** [CaptainHookPhp/captainhook#67]
$ git lfs install
	Hook already exists: pre-push

		#!/usr/bin/env php
		<?php
		$autoLoader = __DIR__ . '/../../vendor/autoload.php';
		
		if (!file_exists($autoLoader)) {
		fwrite(STDERR, 'Composer autoload.php could not be found');
		exit(1);
		}
		require $autoLoader;
		$config = realpath(__DIR__ . '/../../captainhook.json');
		$app    = new CaptainHook\App\Console\Application\Hook();
		$app->setHook('pre-push');
		$app->setConfigFile($config);
		$app->setRepositoryPath(dirname(dirname(__DIR__)));
		$app->run();

	To resolve this, either:
	  1: run `git lfs update --manual` for instructions on how to merge hooks.
	  2: run `git lfs update --force` to overwrite your hook.

So I called 

	$ git lfs update --manual
	Add the following to .git/hooks/pre-push:

		#!/bin/sh
		command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/pre-push.\n"; exit 2; }
		git lfs pre-push "$@"

But how can I merge a shell script with a PHP script?

**Comment by kolaente** [comment]
You could do something like 
```
	#!/bin/sh
echo '	$autoLoader = __DIR__ . '/../../vendor/autoload.php';
	
	if (!file_exists($autoLoader)) {
	fwrite(STDERR, 'Composer autoload.php could not be found');
	exit(1);
	}
	require $autoLoader;
	$config = realpath(__DIR__ . '/../../captainhook.json');
	$app    = new CaptainHook\App\Console\Application\Hook();
	$app->setHook('pre-push');
	$app->setConfigFile($config);
	$app->setRepositoryPath(dirname(dirname(__DIR__)));
	$app->run();' | php -a
	command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/pre-push.\n"; exit 2; }
	git lfs pre-push "$@"

**Comment by amenk** [comment]
Funny idea, but quoting needs to be fixed :-)

**Comment by sebastianfeldmann** [comment]
Or as a workaround do it like this
```
#!/bin/sh
git lfs pre-push "$@"
vendor/bin/captainhook-run "$@"
```

**Comment by sebastianfeldmann** [comment]
Version 5 will provide you the option to install the Cap'n as shell script (instead of a php script now) then you can just append the `LFS` hook at the end

**Comment by sebastianfeldmann** [comment]
This is fixed with `captainhook install --run-mode=shell` but only from version 5.0.0 upwards.

Hopefully version 5 will be released in the next couple of days.

**Comment by amenk** [comment]
do i have to call this manually or is it done during `composer install` in captainhook/plugin-composer ?

**Comment by sebastianfeldmann** [comment]
Sadly you have to do it manually since LFS will not overwrite or append stuff to existing hooks.
I'm working on a feature to make sure `CaptainHook` does not overwrite hooks it installed and that where possibly moddified.
So if you modify it once the Cap'n will not overwrite it even if you use the composer plugin or a `post-autoload-dump` script to make sure the hooks are installed.
In version 5 if the hook is created with the same major version it will not re-install the hook.

Not sure if it could be automated or improved, if you have any ideas ... :)

**Comment by sebastianfeldmann** [comment]
Maybe I could add an option or config setting for `LFS` und append the `pre-push` `LFS` hook to the shell script variant myself.
I would assume the `git lfs pre-push "$@"` is pretty stable and I would not have to update that very often.

**Comment by sebastianfeldmann** [comment]
What also should work is to add the command to the `captainhook.json` as an action.
This has the charm that everybody in the team would have to do it ;)

```json
{
    "pre-push": {
        "enabled": true,
        "actions": [
            {
                "action": "git lfs pre-push '{TARGET}' '{URL}'"
            }
        ]
    }
}
```

**Comment by amenk** [comment]
@sebastianfeldmann What do I have to test? Is it running automatically via the Composer Package now? I cannot identify a commit

**Comment by amenk** [comment]
I tried the following: git lfs was already installed, I required captainhook ^5.0

Now I get (of course)

     pre-push is already installed

Captain hook could do some conflict resultion and add itself after lfs ... but then the question is which of the two, lfs or captainhook should do the logic.

It's a pitty, that something like this is not in git core: https://gist.github.com/mjackson/7e602a7aa357cfe37dadcc016710931b

**Comment by sebastianfeldmann** [comment]
:warning: I updated the placeholder syntax in version 5 if you are still using version you have to use `{NAME}` instead of `{$NAME}`.

What you can do is call the lfs hook commands from captainhook by adding them as actions to the hook configuration. But you have to do it manually (sorry).

For example the `pre-push` hook command
```
{
    "pre-push": {
        "enabled": true,
        "actions": [
            {
                "action": "git lfs pre-push '{$TARGET}' '{$URL}'"
            }
        ]
    }
}
```

I wanted to write an article about this as well, but have a look at this article about husky (a hook manager written in javascript) https://dev.to/mbelsky/pair-husky-with-git-lfs-in-your-javascript-project-2kh0

What the author did was:

- Force install the lfs hooks
- Move the installed hooks from `.git/hooks/*` to `.git/hooks-lfs/*`
- Then install the `CaptainHook` hook install
- Then added the lfs hook to each hook configuration

For the Cap'n this would look like this, remember you have to pass the original hook arguments to the lfs hooks to make it work by using the original argument placeholders `{$NAME}`
```
{
    "pre-push": {
        "enabled": true,
        "actions": [
            {
                "action": ".git/hooks-lfs/pre-push '{$TARGET}' '{$URL}'"
            }
        ]
    }
}
```

I'm still thinking about how to automate this in any way but so far nothing reliable and easy to use came to mind :(

But with this manual solution you should be "fine" :)

**Comment by sebastianfeldmann** [comment]
I just released version 5.1.0 with a new install command option `--move-existing-to=`

So if you install the `git-lfs` hooks first and then run the following:

    vendor/bin/captainhook install --move-existing-to=bin/hooks/

Then the `git-lfs` hooks will be moved to the `bin/hooks` directory. And you will see some hints on how to integrate the moved hooks in your `captainhook.json` file.

I think this is the best I can do for now. Hope that helps a bit.
